Armadillo/src/main/java/com/scribd/armadillo/Reducer.kt

```diff
@@ -66,248 +66,279 @@ internal object Reducer {
-            }
-            is LicenseReleasedAction -> {
-                val oldDrm = oldState.drmPlaybackState
-                return oldState
-                    .copy(drmPlaybackState = DrmState.LicenseReleased(oldDrm.drmType,oldDrm.expireMillis, oldDrm.isSessionValid))
-            }
-            is LicenseKeyIsUsableAction -> {
-                val oldDrm = oldState.drmPlaybackState
-                return oldState
-                    .copy(drmPlaybackState = DrmState.LicenseUsable(oldDrm.drmType, oldDrm.expireMillis))
-            }
-            is LicenseDrmErrorAction -> {
-                val oldDrm = oldState.drmPlaybackState
-                return oldState
-                    .copy(drmPlaybackState = DrmState.LicenseError(oldDrm.drmType, oldDrm.expireMillis))
-            }
+                }
+
+                is OpeningLicenseAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseOpening(action.type ?: oldDrm.drmType))
+                }
+
+                is LicenseAcquiredAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseAcquired(action.type, oldDrm.expireMillis))
+                }
+
+                is LicenseExpiredAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseExpired(oldDrm.drmType, oldDrm.expireMillis))
+                }
+
+                is LicenseExpirationDetermined -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseUsable(oldDrm.drmType, action.expirationMilliseconds))
+                }
+
+                is LicenseReleasedAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseReleased(oldDrm.drmType, oldDrm.expireMillis, oldDrm.isSessionValid))
+                }
+
+                is LicenseKeyIsUsableAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseUsable(oldDrm.drmType, oldDrm.expireMillis))
+                }
+
+                is LicenseDrmErrorAction -> {
+                    val oldDrm = oldState.drmPlaybackState
+                    return oldState
+                        .copy(drmPlaybackState = DrmState.LicenseError(oldDrm.drmType, oldDrm.expireMillis))
+                }
 
-            else -> throw UnrecognizedAction(action)
+                else -> throw UnrecognizedAction(action)
+            }
+        } catch (ex: ActionBeforeSetup) {
```

Comment:
Instead of this, can we handle `ActionBeforeSetup` in the media session? The whole idea behind this was throwing an error if it was called too soon, so it should be on the client to handle it.

Reply:
Imho moving it to the error field of ArmadilloState is the consistent, intended way errors are to be handled.  Clients should be looking into the ArmadilloState for errors, and they shouldn't be expected to look in a 2nd place for them. That's not consistent, and ArmadilloState is a more graceful way to handle these errors without a crash risk.

Reply:
@kabliz LGTM, but wondering why we're not doing this for `UnrecognizedAction` too